vnext/Microsoft.ReactNative/JsiApi.cpp

```diff
@@ -325,15 +328,34 @@ facebook::jsi::JSError const &jsError) {                             \
 /*static*/ ReactNative::JsiRuntime JsiRuntime::GetOrCreate(
     std::shared_ptr<facebook::jsi::RuntimeHolderLazyInit> const &jsiRuntimeHolder,
     std::shared_ptr<facebook::jsi::Runtime> const &jsiRuntime) noexcept {
-  std::scoped_lock lock{s_mutex};
-  auto it = s_jsiRuntimeMap.find(reinterpret_cast<uintptr_t>(jsiRuntime.get()));
-  if (it != s_jsiRuntimeMap.end()) {
-    return it->second.get();
-  } else {
-    ReactNative::JsiRuntime abiJsiResult{make<JsiRuntime>(Mso::Copy(jsiRuntimeHolder), Mso::Copy(jsiRuntime))};
-    s_jsiRuntimeMap.try_emplace(reinterpret_cast<uintptr_t>(jsiRuntime.get()), abiJsiResult);
-    return abiJsiResult;
+  {
+    std::scoped_lock lock{s_mutex};
+    auto it = s_jsiRuntimeMap.find(reinterpret_cast<uintptr_t>(jsiRuntime.get()));
+    if (it != s_jsiRuntimeMap.end()) {
+      return it->second.get();
+    }
   }
+
+  return Create(jsiRuntimeHolder, jsiRuntime);
+}
+
+/*static*/ ReactNative::JsiRuntime JsiRuntime::Create(
+    std::shared_ptr<facebook::jsi::RuntimeHolderLazyInit> const &jsiRuntimeHolder,
+    std::shared_ptr<facebook::jsi::Runtime> const &jsiRuntime) noexcept {
+  auto buffer = std::make_shared<facebook::jsi::StringBuffer>(R"JS(
```

Comment:
Curious, shouldn't JSI be able to modify metadata on functions? I understand what the code does, but the V8 hosting api's and as far as I know chakra hosting api's allow this operation from the native side...

Reply:
I wish it too. JSI is quite limiting in this sense. I do not want to do anything special here: just to create a read-only non-enumerable no-configurable with a Symbol key. This way only my code can access it, while the rest of the system does not see it.

---
In reply to: [537924595](https://github.com/microsoft/react-native-windows/pull/6693#discussion_r537924595) [](ancestors = 537924595)